torch/distributed/fsdp/flat_param.py

```diff
@@ -575,6 +586,404 @@ def shard_metadata(
+                # a GPU tensor (the new sharded gradient).
+                if not grad_offloaded:
+                    flat_param._saved_grad_shard = flat_param.grad.data  # type: ignore[attr-defined]
+            else:
+                padded_unsharded_size = flat_param._full_param_padded.size()  # type: ignore[attr-defined]
+                p_assert(
+                    flat_param.grad.size() == padded_unsharded_size,
+                    "Expects `.grad` to be the unsharded gradient in "
+                    f"`no_sync()` with size {padded_unsharded_size} "
+                    f"but got size {flat_param.grad.size()}",
+                )
+            flat_param.grad = None
+
+    @contextlib.contextmanager
+    def to_cpu(self):
+        """
+        Moves the unpadded unsharded flattened parameter to CPU while in the
+        context and moves it back to the previous device upon exit. For now,
+        this assumes the ``FlatParameter`` is the unpadded unsharded flattened
+        parameter since (1) there is no reason to include the padding in the
+        copy and (2) there is no use case for the sharded flattened parameter.
+
+        Precondition: ``self.flat_param`` 's data is the unpadded unsharded
+            flattened parameter on the compute device, and the handle uses a
+            sharded strategy.
+        Postcondition: Same as the precondition.
+        """
+        self._check_sharded_strategy()
+        p_assert(
+            self.flat_param.size() == self.flat_param._unsharded_size,
+            f"Expects size {self.flat_param._unsharded_size} but got {self.flat_param.size()}",
+        )
+        self._check_on_compute_device(self.flat_param)
+        # Check that the unpadded unsharded flattened parameter is a view into
+        # the padded unsharded flattened parameter as expected
+        # TODO (awgu): For correctness, this check is not strictly needed.
+        # However, the user should not be doing parameter swapping with
+        # `self.flat_param` either.
+        unpadded_storage_ptr = self.flat_param.storage().data_ptr()
+        padded_storage_ptr = (
+            self._get_padded_unsharded_flat_param().storage().data_ptr()
+        )
+        p_assert(
+            unpadded_storage_ptr == padded_storage_ptr,
+            "Expects the unpadded parameter to be a view into the padded parameter",
+        )
+        self._flat_param_to(torch.device("cpu"))
+        self._free_unsharded_flat_param()
+        try:
+            yield
+        finally:
+            p_assert(
+                self.flat_param.size() == self.flat_param._unsharded_size,
+                f"Expects size {self.flat_param._unsharded_size} but got {self.flat_param.size()}",
+            )
+            padded_unsharded_flat_param = self._alloc_padded_unsharded_flat_param()
+            # Copy from CPU to the compute device
+            padded_unsharded_flat_param[: self.flat_param.numel()].copy_(
+                self.flat_param
```

Comment:
should we use non-blocking here?

Reply:
Yes, I think we can because it does not hurt. However, we also do not pin that CPU memory, so there may be no benefit in practice.

To zoom out, this is only used at the moment for `summon_full_params()`, where overlapping the transfer is not critical. The existing implementation did not use `non_blocking=True`.